### PR TITLE
Raise deprecation warnings if non empty `WorkTag` class is used

### PR DESCRIPTION
**Pull request by dalg24**
Drafted our emitting deprecation warnings when user code specifies non empty class as an execution policy `WorkTag` template argument as a tentative fix for #5223 
The intent is to use it as a base for discussion.
```
In file included from <kokkos>/core/src/Kokkos_Core.hpp:57:
In file included from <kokkos>/cmake-build-default/KokkosCore_Config_DeclareBackend.hpp:47:
In file included from <kokkos>/core/src/decl/Kokkos_Declare_THREADS.hpp:49:
In file included from <kokkos>/core/src/Kokkos_Threads.hpp:199:
In file included from <kokkos>/core/src/Kokkos_ExecPolicy.hpp:60:
<kokkos>/core/src/impl/Kokkos_AnalyzePolicy.hpp:224:3: warning: 'warn_if_deprecated' is deprecated: Invalid WorkTag template argument in execution policy!! [-Wdeprecated-declarations]
  KOKKOS_IMPL_STATIC_WARNING(!std::is_empty<typename base_t::work_tag>::value &&
  ^
<kokkos>/core/src/impl/Kokkos_AnalyzePolicy.hpp:216:7: note: expanded from macro 'KOKKOS_IMPL_STATIC_WARNING'
      warn_if_deprecated(std::integral_constant<bool, __VA_ARGS__>()), "")
      ^
<kokkos>/core/src/Kokkos_ExecPolicy.hpp:99:28: note: in instantiation of template class 'Kokkos::Impl::PolicyTraits<Kokkos::HostSpace>' requested here
class RangePolicy : public Impl::PolicyTraits<Properties...> {
                           ^
<kokkos>/core/unit_test/default/TestDefaultDeviceDevelop.cpp:59:30: note: in instantiation of template class 'Kokkos::RangePolicy<Kokkos::HostSpace>' requested here
  static_assert(std::is_same<Policy::work_tag, Kokkos::HostSpace>::value, "");
                             ^
<kokkos>/core/src/impl/Kokkos_AnalyzePolicy.hpp:211:1: note: 'warn_if_deprecated' has been explicitly marked deprecated here
KOKKOS_DEPRECATED_WITH_COMMENT(
^
<kokkos>/core/src/Kokkos_Macros.hpp:610:51: note: expanded from macro 'KOKKOS_DEPRECATED_WITH_COMMENT'
#define KOKKOS_DEPRECATED_WITH_COMMENT(comment) [[deprecated(comment)]]
                                                  ^
1 warning generated.

### CODE REVIEW

File: core/src/impl/Kokkos_AnalyzePolicy.hpp [hunk 1]
```diff
@@ -205,14 +205,27 @@ struct ExecPolicyTraitsWithDefaults : AnalysisResults {
 };
 
 //------------------------------------------------------------------------------
+
+constexpr bool warn_if_deprecated(std::false_type) { return true; }
+KOKKOS_DEPRECATED_WITH_COMMENT(
+    "Invalid WorkTag template argument in execution policy!!")
+constexpr bool warn_if_deprecated(std::true_type) { return true; }
```
Comment (crtrott) [inline core/src/impl/Kokkos_AnalyzePolicy.hpp:212]: shouldn't this return false?

**Review by crtrott — CHANGES_REQUESTED** [review]

**Comment by ajpowelsnl** [comment]
@msimberg would you mind having a look at the possibility of a race condition around `SharedHostPointerTracking`?  We are seeing intermittent test failures.  We are not certain that it's a failure on our side, but it could be.

```
 RUN      ] hpx.host_shared_ptr_tracking
/home/runner/work/kokkos/kokkos/kokkos/core/unit_test/TestHostSharedPtrAccessOnDevice.hpp:208: Failure
Expected equality of these values:
  2
  f1.use_count()
    Which is: 3
[  FAILED  ] hpx.host_shared_ptr_tracking (0 ms)
```

File: core/src/impl/Kokkos_AnalyzePolicy.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/impl/Kokkos_AnalyzePolicy.hpp:212]: Ah so maybe if deprecated code is disabled it should return false ...

**Review by crtrott — CHANGES_REQUESTED** [review]

**Comment by dalg24** [comment]
Failure is one of the CUDA build on a machine running out of disk space
"Safe" to ignore

**Review by crtrott — APPROVED** [review]

File: core/src/traits/Kokkos_WorkTagTrait.hpp [hunk 2]
```diff
@@ -110,6 +110,9 @@ struct WorkTagTrait : TraitSpecificationBase<WorkTagTrait> {
   template <class T>
   using trait_matches_specification = std::integral_constant<
       bool, !std::is_void<T>::value &&
+#ifndef KOKKOS_ENABLE_DEPRECATED_CODE_3
+                std::is_empty<T>::value &&
```
Comment (nliber) [inline core/src/traits/Kokkos_WorkTagTrait.hpp:114]: When the deprecation is removed, we should get rid of the `!std::is_void<T>::value` check, correct?  Maybe leave a comment to that effect, or a `#else` on the macro?

**Review by nliber — CHANGES_REQUESTED** [review]

File: core/src/traits/Kokkos_WorkTagTrait.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/traits/Kokkos_WorkTagTrait.hpp:114]: No, when the deprecation is removed we drop the macro guard that is all.
`void` is not a valid `WorkTag`

Comment (dalg24) [reply core/src/traits/Kokkos_WorkTagTrait.hpp:114]: Gotcha.  Will change.

**Review by masterleinad — CHANGES_REQUESTED** [review]
We should also guard all of the changes in `core/src/impl/Kokkos_AnalyzePolicy.hpp` with `KOKKOS_ENABLE_DEPRECATED_CODE_3`. With the additional condition in `core/src/traits/Kokkos_WorkTagTrait.hpp` it won't really show up in error messages anyway. For me, the main error message then is
```
/tmp/kokkos/core/src/impl/Kokkos_AnalyzePolicy.hpp:172:7: error: implicit instantiation of undefined template 'Kokkos::Impl::show_name_of_invalid_execution_policy_trait<Kokkos::HostSpace>'
      show_name_of_invalid_execution_policy_trait<Trait>{};
      ^
```

**Review by nliber — APPROVED** [review]

**Review by nliber — APPROVED** [review]

**Comment by dalg24** [comment]
> We should also guard all of the changes in `core/src/impl/Kokkos_AnalyzePolicy.hpp` with `KOKKOS_ENABLE_DEPRECATED_CODE_3`. With the additional condition in `core/src/traits/Kokkos_WorkTagTrait.hpp` it won't really show up in error messages anyway. For me, the main error message then is
> 
> ```
> /tmp/kokkos/core/src/impl/Kokkos_AnalyzePolicy.hpp:172:7: error: implicit instantiation of undefined template 'Kokkos::Impl::show_name_of_invalid_execution_policy_trait<Kokkos::HostSpace>'
>       show_name_of_invalid_execution_policy_trait<Trait>{};
>       ^
> ```

Done

**Review by masterleinad — APPROVED** [review]

**Comment by ajpowelsnl** [comment]
What will replace the soon-to-be deprecated non-empty WorkTag class?

**Comment by masterleinad** [comment]
> What will replace the soon-to-be deprecated non-empty WorkTag class?

There is no replacement. Users should just make sure that their tags are empty classes.

**Comment by ajpowelsnl** [comment]
> > What will replace the soon-to-be deprecated non-empty WorkTag class?
> 
> There is no replacement. Users should just make sure that their tags are empty classes.

What was the motivation for this change?

**Pull request merged**